File: 56. Merge Intervals/MergeIntervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    //pay attention!!
    struct comp{
        bool operator()(Interval &i1,Interval &i2){
            return i1.start<i2.start;
        }
    };
    vector<Interval> merge(vector<Interval>& intervals) {
        vector<Interval> res;
        if(intervals.empty() || intervals.size()<=1)
            return intervals;
        sort(intervals.begin(),intervals.end(),comp());
        int len=intervals.size();
        Interval me(-999,-999);
        for(int i=0;i<len;i++)
        {
            if(me.start==-999 && me.end==-999)//me is empty
            {
                me=intervals[i];
            }
            else if(me.end < intervals[i].start){// not overlap. push_back me.
                res.push_back(me);
                me.start=-999;
                me.end=-999;
                i--;
            }
            //else if(me.start> intervals[i].end){// not overlap. push_back interval.
            //    res.push_back(intervals[i]);
            //}
            else// overlap. Merge the two.
            {
                me.start=min(intervals[i].start, me.start);
                me.end= max(intervals[i].end, me.end);
            }
        }
        if(me.start!=-999 || me.end!=-999)
            res.push_back(me);
        
        return res;
    }
};

File: 56. Merge Intervals/MergeIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    private List<Interval> res = new ArrayList<Interval>();
    public List<Interval> merge(List<Interval> intervals) {
        if(intervals == null || intervals.size() == 0)
        {
            return res;
        }
        Collections.sort(intervals, new Comparator<Interval>()
        {
            public int compare(Interval o1, Interval o2)
            {
                if(o1.start == o2.start && o1.end == o2.end)
                    return 0;
                if(o1.start == o2.start)
                {
                    return o1.end < o2.end ? -1: 1;
                }
                else 
                {
                    return o1.start < o2.start? -1: 1;
                }
            }
        });
        Interval current = intervals.get(0);
        for(int i = 1; i < intervals.size(); i++)
        {
            if(current.end >= intervals.get(i).start)
            {
                current.end = intervals.get(i).end < current.end ? current.end : intervals.get(i).end;
            }
            else
            {
                res.add(current);
                current = intervals.get(i);
            }
        }
        res.add(current);
        return res;
    }
}